libgadget/treewalk.c

```diff
@@ -707,24 +632,71 @@ ev_communicate(void * sendbuf, void * recvbuf, size_t elsize, const struct SendR
 }
 
 /* returns the remote particles */
-static void ev_get_remote(const struct SendRecvBuffer sndrcv, TreeWalk * tw)
+static struct SendRecvBuffer ev_get_remote(TreeWalk * tw)
 {
-    size_t j;
+    const int NTask = tw->NTask;
+    size_t i;
     double tstart, tend;
 
+    struct SendRecvBuffer sndrcv = {0};
+    sndrcv.Send_count = (int *) ta_malloc("Send_count", int, 4*NTask+1);
+    sndrcv.Recv_count = sndrcv.Send_count + NTask+1;
+    sndrcv.Send_offset = sndrcv.Send_count + 2*NTask+1;
+    sndrcv.Recv_offset = sndrcv.Send_count + 3*NTask+1;
+
+    /* Fill the communication layouts */
+    /* Use the last element of SendCount to store the partially exported particles.*/
+    memset(sndrcv.Send_count, 0, sizeof(int)*(NTask+1));
+    for(i = 0; i < tw->Nexport; i++) {
+        sndrcv.Send_count[DataIndexTable[i].Task]++;
+    }
+
+    tstart = second();
+    MPI_Alltoall(sndrcv.Send_count, 1, MPI_INT, sndrcv.Recv_count, 1, MPI_INT, MPI_COMM_WORLD);
+    tend = second();
+    tw->timewait1 += timediff(tstart, tend);
+
+    for(i = 0, tw->Nimport = 0, sndrcv.Recv_offset[0] = 0, sndrcv.Send_offset[0] = 0; i < (size_t) NTask; i++)
+    {
+        tw->Nimport += sndrcv.Recv_count[i];
+
+        if(i > 0)
+        {
+            sndrcv.Send_offset[i] = sndrcv.Send_offset[i - 1] + sndrcv.Send_count[i - 1];
+            sndrcv.Recv_offset[i] = sndrcv.Recv_offset[i - 1] + sndrcv.Recv_count[i - 1];
+        }
+    }
+
+    size_t j;
     void * recvbuf = mymalloc("EvDataGet", tw->Nimport * tw->query_type_elsize);
     char * sendbuf = mymalloc("EvDataIn", tw->Nexport * tw->query_type_elsize);
 
     tstart = second();
-    /* prepare particle data for export */
-#pragma omp parallel for
+    int * localsendcount = ta_malloc("localsendcount", int, NTask);
+    memset(localsendcount, 0, sizeof(int)*NTask);
+    /* prepare particle data for export, building the sendbuffer.
+     * Set DataIndexTable.IndexGet to the index in the send buffer. */
     for(j = 0; j < tw->Nexport; j++)
```

Comment:
I think a lot of the pessimization comes from this loop no longer being omp parallel